store/store.go

```diff
@@ -285,116 +291,59 @@ func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
+	blockHashClause := func(key, value []byte, batch dbm.Batch) error {
+		// unmarshal block meta
 		var pbbm = new(tmproto.BlockMeta)
-		bz := iter.Value()
-
-		err = proto.Unmarshal(bz, pbbm)
+		err := proto.Unmarshal(value, pbbm)
 		if err != nil {
-			panic(fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err))
+			return fmt.Errorf("unmarshal to tmproto.BlockMeta: %w", err)
 		}
 
 		blockMeta, err := types.BlockMetaFromProto(pbbm)
 		if err != nil {
-			panic(fmt.Errorf("error from proto blockMeta: %w", err))
+			return fmt.Errorf("error from proto blockMeta: %w", err)
 		}
 
+		// delete the hash key corresponding to the block meta's hash
 		if err := batch.Delete(blockHashKey(blockMeta.BlockID.Hash)); err != nil {
-			return 0, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
+			return fmt.Errorf("failed to delete hash key: %X: %w", blockHashKey(blockMeta.BlockID.Hash), err)
 		}
 
-		pruned++
-		// avoid batches growing too large by flushing to database regularly
-		if pruned%1000 == 0 {
-			if err := iter.Error(); err != nil {
-				return pruned, err
-			}
-			if err := iter.Close(); err != nil {
-				return pruned, err
-			}
+		return nil
+	}
 
-			err := batch.Write()
-			if err != nil {
-				return pruned, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
-			}
-			if err := batch.Close(); err != nil {
-				return pruned, err
-			}
+	// remove block meta first as this is used to indicate whether the block exists.
+	// For this reason, we also use ony block meta as a measure of the amount of blocks pruned
+	pruned, err := bs.batchDelete(blockMetaKey(0), blockMetaKey(height), blockHashClause)
+	if err != nil {
+		return pruned, err
+	}
 
-			iter, err = bs.db.Iterator(
-				blockMetaKey(1),
-				blockMetaKey(retainHeight),
-			)
-			if err != nil {
-				panic(err)
-			}
-			defer iter.Close()
+	// for the rest of the block data we just want to remove it
+	emptyClause := func(key, value []byte, batch dbm.Batch) error { return nil }
 
-			batch = bs.db.NewBatch()
-			defer batch.Close()
-		} else {
-			iter.Next()
-		}
+	if _, err := bs.batchDelete(blockPartKey(0, 0), blockPartKey(height, 0), emptyClause); err != nil {
+		return pruned, err
 	}
-	if err := iter.Error(); err != nil {
+
+	if _, err := bs.batchDelete(blockCommitKey(0), blockCommitKey(height), emptyClause); err != nil {
 		return pruned, err
 	}
 
-	err = batch.WriteSync()
-	if err != nil {
-		return pruned, fmt.Errorf("pruning error at key %X: %w", iter.Key(), err)
+	if _, err := bs.batchDelete(seenCommitKey(0), seenCommitKey(height), emptyClause); err != nil {
+		return pruned, err
 	}
 
 	return pruned, nil
 }
 
 // batchDelete is a generic function for deleting a range of values based on the lowest
 // height up to but excluding retainHeight
-func (bs *BlockStore) batchDelete(start []byte, end []byte) error {
+func (bs *BlockStore) batchDelete(
+	start []byte,
+	end []byte,
+	clause func(key, value []byte, batch dbm.Batch) error) (uint64, error) {
```

Comment:
The term `clause` isn't immediately obvious to me -- what does it mean in this context? We should explain it in the comment as well.

Also, maybe move `(uint64, error)` to the next line, I found it a bit hard to parse this line since there's two different return value lists here.

Reply:
Yeah I also wasn't too sure if it explained what it was for. I was thinking that clause represented a secondary or auxiliary step in deletion. Perhaps something like `preprocess` might be more apt. What do you think?

Reply:
Maybe just `process` -- preprocess sort of indicates that the results will be used as input to some other operation. Not really a great term either, but don't have any better ideas.

Reply:
Basically this is a function that is called on a key/value pair before we do the deletion hence why I thought of the `pre` part. Perhaps `preDeletionHook`

Reply:
Sure